caffe2/onnx/backend.cc

```diff
@@ -674,25 +674,40 @@ Caffe2Ops Caffe2Backend::CreateGemm(
 
   auto trans_a = onnx_node->attributes.get<int64_t>("transA", 0L);
   auto trans_b = onnx_node->attributes.get<int64_t>("transB", 0L);
-  auto broadcast = onnx_node->attributes.get<int64_t>("broadcast", 0L);
+  // Support broadcast by default when opset_version > 6.
+  auto broadcast =
+    onnx_node->attributes.get<int64_t>("broadcast",
+                                       (ctx.opset_version() > 6) ? 1L : 0L);
 
   bool use_fc = false;
-  if ((!trans_a) && trans_b) {
-    if (broadcast) {
-      use_fc = true;
-    } else {
-      const auto input_c_vi_iter = ctx.value_infos().find(node.input(2));
-      if (input_c_vi_iter != ctx.value_infos().end() &&
-          input_c_vi_iter->second.type().tensor_type().shape().dim_size() ==
-              1) {
-        use_fc = true;
+  if (!trans_a && broadcast) {
+    // If we have broadcast and the input_c's dim matching the FC requirement,
+    // we could convert this onnx op to caffe2 FC op.
+    const auto input_c_vi_iter = ctx.value_infos().find(node.input(2));
+    const auto input_b_vi_iter = ctx.value_infos().find(node.input(1));
+    if (input_c_vi_iter != ctx.value_infos().end() &&
+        input_c_vi_iter->second.type().tensor_type().has_shape() &&
+        input_b_vi_iter != ctx.value_infos().end() &&
+        input_b_vi_iter->second.type().tensor_type().has_shape()) {
+      const auto input_c_shape = input_c_vi_iter->second.type().tensor_type().shape();
+      const auto input_b_shape = input_b_vi_iter->second.type().tensor_type().shape();
+      int input_b_last_dim_index = (trans_b) ? 0 : 1;
+      if (input_c_shape.dim_size()==1 && input_c_shape.dim(0).has_dim_value() &&
+          input_b_shape.dim_size()==2 && input_b_shape.dim(input_b_last_dim_index).has_dim_value()) {
+        if (input_c_shape.dim(0).dim_value()==input_b_shape.dim(input_b_last_dim_index).dim_value()) {
```

Comment:
If this check doesn't pass, it means the onnx model is *invalid*, so it's not just a matter of whether using fc or not, instead here should raise error.

Reply:
It's possible... if c is one-element tensor.

Reply:
@houseroad in that case, explicitly check that

Reply:
Very legit. @JerryShih could you address the above comments?

Reply:
I still can't catch the idea. If there is no shape info, we can't make sure the input c fit the FC op's requirement(just like the scalar c case in previous comment). In this case, I think we should just use "MatMul+add". 
What did I miss?

@bddppq @houseroad

Reply:
If only C's shape information is available and C is a 1d tensor, but B's shape is not available, we can aggressively assume that C's shape matches B's last dimension (or first dimension if transB). 1-element tensor is special, you can handle it with `MatMul + Add`. The point here is that `FC`'s performance is much better than `MatMul + Add`, so we would like to use `FC` more aggressively.